File: src/main/java/com/example/chatService/kafka/DMStoreConsumer.java
package com.example.chatService.kafka;

import com.example.chatService.dto.DMMessageKafkaDto;
import com.example.chatService.entity.DMMessage;
import com.example.chatService.entity.DMRoom;
import com.example.chatService.repository.DMMessageRepository;
import com.example.chatService.repository.DMRoomRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneId;

@Service
@RequiredArgsConstructor
public class DMStoreConsumer {

    private final DMRoomRepository roomRepository;
    private final DMMessageRepository messageRepository;

    @KafkaListener(
            topics = "dm-messages",
            groupId = "dm-store", // 공통 groupId
            containerFactory = "DMKafkaListenerContainerFactory"
    )
    public void store(DMMessageKafkaDto dto) {

        DMRoom room = roomRepository.findById(dto.getRoomId())
                .orElseThrow(() -> new RuntimeException("Room not found"));

        OffsetDateTime sentAt =
                OffsetDateTime.ofInstant(
                        Instant.ofEpochMilli(dto.getSentAt()),
                        ZoneId.of("Asia/Seoul")
                );

        DMMessage message = DMMessage.builder()
                .room(room)
                .senderId(dto.getSenderId())
                .content(dto.getContent())
                .sentAt(sentAt)
                .isRead(false)
                .build();

        messageRepository.save(message);
    }
}

File: src/main/java/com/example/chatService/kafka/DmStoreConsumer.kt
package com.example.chatService.kafka

import com.example.chatService.dto.DMMessageKafkaDto
import com.example.chatService.entity.DMMessage
import com.example.chatService.entity.DMRoom
import com.example.chatService.repository.DMMessageRepository
import com.example.chatService.repository.DMRoomRepository
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.stereotype.Service
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneId

@Service
class DmStoreConsumer (
        private val roomRepository : DMRoomRepository,
        private val messageRepository : DMMessageRepository
){
    @KafkaListener(
            topics = ["dm-messages"],
            groupId = "dm-store",
            containerFactory = "DMKafkaListenerContainerFactory"
    )
    fun store(dto: DMMessageKafkaDto) {
        val room : DMRoom = roomRepository.findById(dto.roomId)
                .orElseThrow { RuntimeException("Room not found") }

        val sentAt = OffsetDateTime.ofInstant(
                Instant.ofEpochMilli(dto.sentAt),
                ZoneId.of("Asia/Seoul")
        )

        val message = DMMessage(
                room = room,
                senderId = dto.senderId,
                senderName = dto.senderName,
                content = dto.content,
                sentAt = sentAt,
                isRead = false
        )
        messageRepository.save(message)
    }
}